### PR TITLE
XSS Vulnerabilty fix on linked object displayname/title

### PR DESCRIPTION
**Pull request by prangonswachha-okta**
This pull request addresses a recently identified cross-site scripting (XSS) vulnerability within rockstar client extension. The vulnerability was reported by a third party researcher that on link objects displayname if there is any cross site scripting payload then the extension renders the payload executing the script. 

For example for the payload like when set as the display name for a linked object in directory and "Show Linked Object" is triggered-
`Manager`

### CODE REVIEW

File: rockstar/dompurify-3.2.3.min.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+/*! @license DOMPurify 3.2.3 | (c) Cure53 and other contributors | Released under the Apache license 2.0 and Mozilla Public License 2.0 | github.com/cure53/DOMPurify/blob/3.2.3/LICENSE */
```
Comment (prangonswachha-okta) [inline rockstar/dompurify-3.2.3.min.js:1]: Sourced the file from Dompurify's repo https://github.com/cure53/DOMPurify/blob/ad0fe79863f24d9a6e6583a350e2efb791f77bc9/dist/purify.min.js#L1

File: rockstar/rockstar.js [hunk 2]
```diff
@@ -449,15 +449,18 @@
             }
             async function getLink(lo) {
                 const div = loPopup[0].appendChild(document.createElement('div'));
-                div.innerHTML = lo.title + '<br>Loading...<br><br>';
+                const sanitizedTitle = DOMPurify.sanitize(lo.title);
```
Comment (prangonswachha-okta) [inline rockstar/rockstar.js:452]: Sanitized the linked object title from xss here.

**Comment by gabrielsroka** [comment]
thanks for bringing it to my attention. is there a way to do this without dompurify? eg, i have an escape function `e(s)`

**Comment by gabrielsroka** [comment]
or even just use .textContent/.innerText instead of .innerHTML (for lo.title)

**Comment by prangonswachha-okta** [comment]
> thanks for bringing it to my attention. is there a way to do this without dompurify? eg, i have an escape function `e(s)`

Thanks @gabrielsroka ! So you can use any of them to be honest. 

- If you use escape function just the lo title will escape any special character.

- You can just strip out the html characters and tag and/or keep alphanumeric char e.g.
`function stripAllTags(html) {
    return html.replace(/<[^>]*>/g, '');
}
`
- innerText works! Example.
`const element = document.getElementById('safeElement');
element.textContent = '<script>alert("XSS")</script>'; // This will display as text`

The reason I used Dompurify as its kind of an industry standard and many custom sanitizer has their weakness, if the library becomes vulnerable in the future there is public awareness to it. For this case though just stripping out all dangerous character is a valid option as the linked object display name isn't a breaking change. Will leave it to your best judgement. The only thing is as long as you don't manually append unsanitized data to html we should be fine.

**Comment by gabrielsroka** [comment]
ok. btw, i tried creating a linked object (via the UI) with `<script>alert(1)</script>` and it wouln't let me.

do u have an example of this a) actually happening, or b) theoretically happening?

**Comment by gabrielsroka** [comment]
i tried again, it did allow me to create a linked object, but rockstar didn't do anything (i think the CSP on the Okta page stopped it)

**Comment by prangonswachha-okta** [comment]
> ok. btw, i tried creating a linked object (via the UI) with `<script>alert(1)</script>` and it wouln't let me.
> 
> do u have an example of this a) actually happening, or b) theoretically happening?

Example payload - 

`Manager`

Note that it was working a while ago we added some protection that prevents inline scipt but its best to have defence in depth now. Following screenshots are below from the POC.

Screenshot 2024-11-14 at 4 09 55 PM
Screenshot 2024-11-14 at 4 27 09 PM

**Comment by gabrielsroka** [comment]
interesting. i tried ur value, it tried to fetch , fails and doesn't do the onerror.

i agree about defense in depth, but i also want to know if this is an actual or a theoretical issue

**Comment by gabrielsroka** [comment]
also, if Okta wants to do defense in depth, why doesn't it check for values like this? i asked the same question years ago and got nothing

**Comment by prangonswachha-okta** [comment]
> interesting. i tried ur value, it tried to fetch ``, fails and doesn't do the onerror.
> 
> i agree about defense in depth, but i also want to know if this is an actual or a theoretical issue

It was an working exploit after the added security changes recently its no longer. Please refer to the comment above I have added the screenshot from when it made those fetch request to change password.

**Comment by gabrielsroka** [comment]
ok. i just wanna know how to repro. i can't

**Comment by gabrielsroka** [comment]
ok. i updated the code. i'll deploy to Google asap.

**Pull request closed without merging**